gtsam/slam/SmartProjectionRigFactor.h

```diff
@@ -0,0 +1,366 @@
+
+  static const int DimPose = 6;  ///< Pose3 dimension
+  static const int ZDim = 2;  ///< Measurement dimension
+
+ protected:
+
+  /// vector of keys (one for each observation) with potentially repeated keys
+  KeyVector nonUniqueKeys_;
+
+  /// cameras in the rig (fixed poses wrt body + fixed intrinsics, for each camera)
+  typename Base::Cameras cameraRig_;
+
+  /// vector of camera Ids (one for each observation, in the same order), identifying which camera took the measurement
+  FastVector<size_t> cameraIds_;
+
+ public:
+  typedef CAMERA Camera;
+  typedef CameraSet<CAMERA> Cameras;
+
+  /// shorthand for a smart pointer to a factor
+  typedef boost::shared_ptr<This> shared_ptr;
+
+  /// Default constructor, only for serialization
+  SmartProjectionRigFactor() {
+  }
+
+  /**
+   * Constructor
+   * @param sharedNoiseModel isotropic noise model for the 2D feature measurements
+   * @param cameraRig set of cameras (fixed poses wrt body and intrinsics) in the camera rig
+   * @param params parameters for the smart projection factors
+   */
+  SmartProjectionRigFactor(const SharedNoiseModel& sharedNoiseModel,
+                           const Cameras& cameraRig,
+                           const SmartProjectionParams& params =
+                               SmartProjectionParams())
+      : Base(sharedNoiseModel, params),
+        cameraRig_(cameraRig) {
+    // use only configuration that works with this factor
+    Base::params_.degeneracyMode = gtsam::ZERO_ON_DEGENERACY;
+    Base::params_.linearizationMode = gtsam::HESSIAN;
+  }
+
+  /**
+   * Constructor
+   * @param sharedNoiseModel isotropic noise model for the 2D feature measurements
+   * @param camera single camera (fixed poses wrt body and intrinsics)
+   * @param params parameters for the smart projection factors
+   */
+  SmartProjectionRigFactor(const SharedNoiseModel& sharedNoiseModel,
+                           const Camera& camera,
+                           const SmartProjectionParams& params =
+                               SmartProjectionParams())
+      : Base(sharedNoiseModel, params) {
+    // use only configuration that works with this factor
+    Base::params_.degeneracyMode = gtsam::ZERO_ON_DEGENERACY;
+    Base::params_.linearizationMode = gtsam::HESSIAN;
+    cameraRig_.push_back(camera);
+  }
+
+  /** Virtual destructor */
+  ~SmartProjectionRigFactor() override {
+  }
+
+  /**
+   * add a new measurement, corresponding to an observation from pose "poseKey"
+   * and taken from the camera in the rig identified by "cameraId"
+   * @param measured 2-dimensional location of the projection of a
+   * single landmark in a single view (the measurement)
+   * @param poseKey key corresponding to the body pose of the camera taking the measurement
+   * @param cameraId ID of the camera in the rig taking the measurement (default 0)
+   */
+  void add(const Point2& measured, const Key& poseKey, const size_t cameraId = 0) {
```

Comment:
Should be `const MEASUREMENT& measured` so that other camera types can be used. With this it works with #861

Reply:
I'll apply this in #861. can you add this comment there?

Reply:
scratch that: the "Measurement" infrastructure is only introduced in #861 , so this does not compile here. I'll need to apply your suggestions after I merge.